moves functionality to Python

yawpitch opened the pull request:
Eliminates all but a bash shim file, restructures more in line with the python-representer and python-anaylzer repos.

To run just `sh bin/run.sh SLUG EXERCISE_DIR OUTPUT_DIR.

Early days: doesn't work in Docker (yet) and I've just hacked together what the "message" should be. Could use some definite refactoring too, once we've got a better sense of the actually required output.

cmccandless commented:
It might be time to un-mark WIP on this PR so that GH actions can run and validate the changes here.

yawpitch commented:
So that seems to have triggered a build ... but I can't actually access it, and it shouldn't really _still_ be running.

cmccandless commented:
I can't access it either... Try triggering with an empty commit? I'm assuming this will be squashed on merge anyway.

yawpitch commented:
Meh ... I'll do more work on this today and hopefully trigger it again soon.

yawpitch commented:
@cmccandless the build isn't working because it's on my fork, and I haven't enabled ... which is fine, because I still haven't touched the docker approach on this.

_Massive_ update ... got the self test running (just go to the root dir and run `pytest`. I've also added some more example tests, updated to capture stdout, and demonstrated some of the various traceback styles available.

@iHiD if you take a glance at the various results.[style option].json files, all prefixed with `traceback-styles` you'll get a feel for the various outputs Pytest can create.  Be nice to choose on and then sanitize what we have to out of it rather than try for something completely new.

The root level message (for instance in the `example-syntax-error` results) is currently much more verbose, incorporating the full traceback.

One thing I didn't do is add a test for the 500 character truncation... I'll add that in a second.

yawpitch commented:
BTW I went with golden files to save our ability to iterate without pulling out our own hair. If we make a change we just need to run the test files pointed at their directories and update the results.jsons to reflect the new golden state.

cmccandless commented:
> which is fine, because I still haven't touched the docker approach on this.
I had added a commit (https://github.com/exercism/python-test-runner/pull/27/commits/7f2eb13899c191db265cc14112f37ee5f548d301) to make the Dockerfile compatible with your changes. If there haven't been significant directory structure changes since, it should work.

iHiD commented:
@yawpitch So I feel like `results.no.json` is right for assertion fails. And `results.native.json` is right for exceptions (either exceptions in tests or syntax errors). 

What do you think?

@ErikSchierboom Thoughts?

yawpitch commented:
@iHiD personally I feel like the `short` style strikes the right balance. In Python the assertion message is only _very_ rarely of much use ... it tells you _what_ happened, but not where. It's useful for seeing where in the _test_ the problem occurred, not where in your code.

I haven't yet discovered how (or if) I can use two different styles on the same invocation (the style choice is given to pytest _before_ the run begins), but I'll see.

That said the "problem" with the native mode is that it incorporates a lot of system-local noise into the traceback, which means that the golden test approach won't work from machine to machine if I can't sanitize that out.

cmccandless commented on test/traceback-styles/results.native.json:
```diff
@@ -0,0 +1,10 @@
+    {
+      "name": "TracebackStylesTest.test_hello",
+      "status": "fail",
+      "message": "Traceback (most recent call last):\n  File \"/home/michael/Documents/dev/python-test-runner/test/traceback-styles/traceback_styles_test.py\", line 9, in test_hello\n    self.assertEqual(hello(), \"Goodbye\")\n  File \"/usr/lib/python3.8/unittest/case.py\", line 912, in assertEqual\n    assertion_func(first, second, msg=msg)\n  File \"/usr/lib/python3.8/unittest/case.py\", line 1 ...
```
Is the path here going to be a problem for the unit tests?

yawpitch replied:
Definitely, working on fixing that.

cmccandless commented:
@yawpitch Can you please enable GH actions in your fork so we can verify that this runs correctly in a clean environment?

iHiD commented:
I think it would be good for you to both try the Ruby stuff on research.exercism.io and see if that feels good to you in terms of the different messages that appear. It wasn't until I actually played with it live that I found myself understanding what felt best for Ruby.

The username/password are in the #experiment-1 channel on slack.

yawpitch commented:
> @yawpitch Can you please enable GH actions in your fork so we can verify that this runs correctly in a clean environment?

@cmccandless they are active, though because of the local paths they're not working yet. Not sure why Github's not picking the failures up though.

ErikSchierboom commented:
I quite like the `results.no.json` format. Short and to the point.

yawpitch commented:
Ok, just pushed some more changes.

I've eliminated the various "styles" examples, as we're going with the "no" style ... there's still plenty of information for the student if there's a SyntaxError or other "simple" fail, and if there were something more serious we'd have to go to the runner logs anyway.

This took some semi-ludicrous hacking into what pytest was doing ... lord help us if they ever significantly change their API.

@cmccandless I still have no idea why the Action isn't picking up over here, but if you can read the Actions on my fork it's passing in this latest form ... which tends to imply the new Dockerfile works.

I did move the `run_in_docker` script to `bin/` ... I haven't been able to test that as my local install of Docker seems to have gotten royally borked.

cmccandless commented:
@iHiD Since GH isn't picking up on the Action runs from the fork, there's no way for me to merge this under current settings. We might need to consider making the Action run a non-required check until the GH team fixes this bug.

yawpitch commented:
Really very annoying ... a bit of searching on the forums seems say there may be an issue with some outdated checks in the "protected branch" settings, but yeah neither of us has sufficient permissions to try and fix that.

iHiD commented:
I've turned it off for you. Let me know if you ever want anything merging. I can override rules with my org-admin permissions.

yawpitch commented:
Thanks @iHiD!

Right, we're merged, and the Action passed just as it did on my fork, so you're good to deploy!
